abmarl/external/open_spiel_env_wrapper.py

```diff
@@ -0,0 +1,302 @@
+
+    @current_player.setter
+    def current_player(self, value):
+        assert value in self._learning_agents, "Current player must be an agent in the simulation."
+        self._current_player = value
+
+    def reset(self, **kwargs):
+        """
+        Reset the simulation.
+        """
+        self._should_reset = False
+        obs = self.sim.reset(**kwargs)
+        self.current_player = next(iter(obs))
+
+        # If it is a turn based sim, then the current player should be the next
+        # player whose turn it will be. But what is current player for an all-step game?
+        observations = {
+            "info_state": self._append_obs(obs),
+            "legal_actions": {
+                agent_id: self.get_legal_actions(agent_id)
+                for agent_id in self._learning_agents
+            },
+            "current_player": self.current_player,
+        }
+
+        return TimeStep(
+            observations=observations,
+            rewards=None,
+            discounts=None,
+            step_type=StepType.FIRST
+        )
+
+    def step(self, action_list, **kwargs):
+        """
+        Step the simulation forward using the reported actions.
+
+        OpenSpiel provides an action list of either (1) the agent whose turn it
+        is in a turn-based game or (2) all the agents in a simultaneous game. The
+        OpenSpielWrapper converts the list of actions to a dictionary before passing
+        it to the underlying simulation.
+
+        OpenSpiel does not support the ability for some agents of a game to finish
+        before others. As such, it may provide actions for agents that are already
+        done. To work with Abmarl, this wrapper removes actions for agents that
+        are already done.
+        """
+        # Actions come in as a list, so we need to convert to a dict before forwarding
+        # to the SimulationManager.
+        if self.is_turn_based:
+            action_dict = {self.current_player: action_list[0]}
+        else:
+            action_dict = {
+                agent_id: action_list[i]
+                for i, agent_id in enumerate(self._learning_agents)
+            }
+        # OpenSpiel can send actions for agents that are already done, which doesn't
+        # work with our simulation managers. So we filter out these actions before
+        # passing them to the manager.
+        # TODO: Although all implemeted managers do track the done agents, this is not
+        # a part of the SimulationManager interface.
+        for agent_id in self.sim.done_agents:
+            try:
+                del action_dict[agent_id]
+            except KeyError:
+                pass
+        # We have just deleted actions from agents that are already done, which
+        # can result in an empty action dictionary (e.g. in a turn-based game).
+        # In this case, we just take a fake step.
+        if not action_dict: # No actions
+            return self._take_fake_step()
```

Comment:
Add a test in the test suite with an empty dictionary to check that fake stepping works